Explicitly mark declared member initializations in IOperation

JohnHamby opened the pull request:
As requested by @lgolding in #7293, this change set adds the ability to explicitly identify declared initializations of fields in IOperations. The new operation kinds FieldInitializerAtDeclaration, PropertyInitializerAtDeclaration, and ParameterInitializerAtDeclaration denote "= Value" expressions in declarations. Existing operation kinds for field and property initialization in object creation have been renamed to FieldInitializerInCreation and PropertyInitializerInCreation. Both FieldInitializerAtDeclaration and FieldInitiailizerInCreation imply that a node implements IFieldInitializer, and both PropertyInitializerAtDeclaration and PropertyInitializerInCreation imply that a node implements IPropertyInitializer.

@lgolding @genlu @AlekseyTs @nguerrera @srivatsn @CyrusNajmabadi @cston could surely provide helpful feedback.

JohnHamby commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -36,14 +36,37 @@
-  <Node Name="BoundEqualsValue" Base="BoundNode">
-    <!-- Expression representing the value. -->
-    <Field Name="Value" Type="BoundExpression"/>
+  <Node Name="BoundFieldEqualsValue" Base="BoundEqualsValue">
```
I added three derived node types to indicate the kind of initialized symbol. This could be done by instead adding a field of type ISymbol to BoundEqualsValue, should that be preferable.

JohnHamby commented on src/Compilers/Core/CodeAnalysisTest/Diagnostics/CouldHaveMoreSpecificTypeAnalyzer.cs:
```diff
@@ -113,7 +113,21 @@ public sealed override void Initialize(AnalysisContext context)
+                                    {
+                                        // Visual Basic "As New" field declarations don't produce field initializations--they just produce object creation expressions.
+                                        // This is a bug in the VB semantic model support, and requires a subtle fix. For now, code around the issue in this analyzer.
+                                        IExpression expression = operationBlockContext.OperationBlocks[0] as IExpression;
```
This code shouldn't need to be here. I have a question out to @AlekseyTs about how best to cope with the semantic model issue.

JohnHamby commented on src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb:
```diff
@@ -1382,6 +1370,105 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+    Partial Class BoundFieldOrPropertyInitializer
+        Implements IFieldInitializer, IPropertyInitializer
+
+        Private ReadOnly Property IField As IFieldSymbol Implements IFieldInitializer.Field
```
I haven't yet done the editing pass to switch to an InterfaceName_MemberName naming pattern.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -36,14 +36,37 @@
+    <Field Name="Parameter" Type="ParameterSymbol"/>
+  </Node>
+
+  <AbstractNode Name="BoundEqualsValue" Base="BoundNode">
```
Usually base class precedes derived ones.

JohnHamby replied:
Will change. Is this an implied endorsement of having three derived types (for field, property, and parameter declarations) rather than one type that covers all of them?

AlekseyTs replied:
I don't have any objections against three derived types. Given that we want them to implement different interfaces, one type doesn't look like a viable alternative.

cston commented on src/Compilers/Core/Portable/Compilation/IExpression.cs:
```diff
@@ -795,31 +795,21 @@ public interface IObjectCreationExpression : IExpression
     /// Represents an object member initializer.
     /// </summary>
-    public interface IMemberInitializer : IOperation
+    public interface ISymbolInitializer : IOperation
```
Is it necessary to have `ISymbolInitializer` to unify parameter default values and member initialization or should `IMemberInitializer` and `IParameterInitializer` be independent? (Are there cases where an analyzer would want to treat parameter default values and member initializers similarly?)

JohnHamby replied:
It just make the implementations a bit simpler. The sharing is probably not of interest to consumers.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/Expression.cs:
```diff
@@ -902,6 +898,40 @@ partial class BoundConditionalAccess : IConditionalAccessExpression
         protected override OperationKind ExpressionKind => OperationKind.ConditionalAccessExpression;
     }
 
+    partial class BoundEqualsValue : ISymbolInitializer
```
It feels like this file contains to many type declarations. We should consider moving all IOperation- specific code into a separate file(s). One file per type under an IOperation folder would be preferred. This doesn't have to be done as part of this change, but sooner is better than later.

cston commented on src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs:
```diff
@@ -175,15 +175,15 @@ internal override BoundNode Bind(Binder binder, CSharpSyntaxNode node, Diagnosti
         }
 
-        private BoundEqualsValue BindEqualsValue(Binder binder, EqualsValueClauseSyntax equalsValue, DiagnosticBag diagnostics)
+        private BoundNode BindEqualsValue(Binder binder, EqualsValueClauseSyntax equalsValue, DiagnosticBag diagnostics)
```
Is the return type change necessary?

JohnHamby replied:
Probably not. Good catch. It was necessary at some interim point before I (re)introduced BoundEqualsValue as a shared base type.

genlu replied:
Why can't it be `BoundEqualsValue`?

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/IExpression.cs:
```diff
@@ -795,31 +795,21 @@ public interface IObjectCreationExpression : IExpression
     /// Represents an initialization of a field.
     /// </summary>
-    public interface IFieldInitializer : IMemberInitializer
+    public interface IFieldInitializer : ISymbolInitializer
```
Multiple fields can be initialized with one [As New] clause in VB. What is the plan to represent this in the API?

JohnHamby replied:
I have wrestled with this question for a few days. The choices, as I understand them:
  1) Leave the API as is, with only a property that returns the first symbol. This is completely correct for C#, and works correctly for the vast majority of VB programs. However, for VB programs that initialize multiple fields with As New or array types with sizes, the behavior is incorrect with no workaround (except for falling back on operation block actions).
  2) Have both a property that returns the first symbol and one that returns all symbols. In practice, this is as bad as 1) because almost nobody will write analyzers that examine all the symbols and the additional property will be confusing.
  3) Have only a property that returns all the symbols. This will confuse and annoy almost all analyzer authors, but does make it more likely that analyzers will be correct.

3) is not satisfying but is the only defensible choice.

Anyone care to argue a different position?

JohnHamby replied:
Within the VB compiler, there is logic for handling multiple properties that share an initializer in addition to multiple fields that share an initializer. Is it actually possible to write VB code in which two properties share an initializer?

genlu replied:
Among those 4 options, I'd settle for (3) with detailed comments to avoid confusion.

JohnHamby replied:
OK, I'm about to push a new commit that replaces the InitializedField property with an InitializedFields property. This makes it possible to encode all of the fields declared with an As New declaration. However, it doesn't work end to end because MemberSemanticModel and InitializerSemanticModel conspire to bake in an assumption that an initializer applies to precisely one symbol. I'm not going to fix this as part of this change set, but I can open an issue for it.

Also, my assumption is that there is no way for a property initializer to apply to more than one property, despite at least some of the VB compiler's implementation being prepared for that. @AlekseyTs, is this assumption correct?

AlekseyTs replied:
I think the assumption is correct.

lgolding replied:
@michaelcfanning FYI. You might find it amusing or interesting that the issue they wrestled with here ("there's almost always only one of something, but we _have_ to support the rare case where there can be two or more of them") is the same one we keep encountering in defining the SARIF format. And they came to the same conclusion we did: "Always represent the thing as a collection, even though it will almost always have only one element."

cston commented on src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb:
```diff
@@ -134,7 +134,21 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-            Return If(expressionInitializer IsNot Nothing, New BoundEqualsValue(initializer, expressionInitializer), boundInitializer)
+            If expressionInitializer IsNot Nothing Then
+                If Me.MemberSymbol.Kind = SymbolKind.Parameter Then
+                    Return New BoundParameterEqualsValue(initializer, DirectCast(Me.RootBinder.ContainingMember, ParameterSymbol), expressionInitializer)
```
`Me.MemberSymbol` rather than `Me.RootBinder.ContainingMember`?

cston replied:
Consider changing `Ifs` to `Select Case Me.MemberSymbol.Kind`.

JohnHamby replied:
Of right, I forgot to call this out. Preexisting code above gets to the parameter symbol via Me.RootBinder.ContainingMember, and so I copied that. @AlekseyTs, why does the code above do that rather than access Me.MemberSymbol?

AlekseyTs replied:
I don't know, @mavasani might know.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb:
```diff
@@ -134,7 +134,21 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                End If
+
+                If Me.MemberSymbol.Kind = SymbolKind.Property Then
+                    Return New BoundPropertyEqualsValue(initializer, DirectCast(Me.MemberSymbol, PropertySymbol), expressionInitializer)
```
Do we need to dig through BoundFieldOrPropertyInitializer as GetOperationWorker used to do?

JohnHamby replied:
No. BoundFieldOrPropertyInitializer is not an expression, and one of those serves just as well as a Bound(Field/Property)EqualsValue. See the related discussion below.

cston commented on src/Compilers/CSharp/Portable/BoundTree/Expression.cs:
```diff
@@ -902,6 +898,40 @@ partial class BoundConditionalAccess : IConditionalAccessExpression
+
+    partial class BoundPropertyEqualsValue : IPropertyInitializer
+    {
+        IMethodSymbol IPropertyInitializer.Setter => this.Property?.SetMethod;
```
When is `this.Property` null?

cston replied:
`IPropertyInitializer` should probably expose `Property` rather than `Setter`. Otherwise, in an error case where the property is readonly, the analyzer will not be able to determine the property.

genlu replied:
Agreed, @AlekseyTs brought this up in one of the previous `IOperation` PRs as well.

cston replied:
Ah, I thought I'd seen a similar comment earlier, thanks.

JohnHamby replied:
Having the setter would be important if it were possible to overload property setters. However, it's not, so I'll change the API to return the property symbol instead.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb:
```diff
@@ -1382,6 +1370,105 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
     End Class
 
+    Partial Class BoundFieldOrPropertyInitializer
+        Implements IFieldInitializer, IPropertyInitializer
```
I think implementing two interfaces on this node could be confusing to consumers. I am not sure why do we expose this node at all since we added BoundPropertyEqualsValue and BoundFieldEqualsValue.

JohnHamby replied:
BoundFieldOrPropertyInitializer gets generated in most but apparently not all circumstances. This node is generated, I think, for normal compilation as well as for semantic models, so I am reluctant to try to figure out how to elide them.

If used in the expected pattern--check Kind and then cast based on Kind--things will work out OK. If used by testing for type rather than testing for Kind, then confusion is likely.

Given that this node type is used by the compiler, it's probably simpler to use it for all = Value constructs rather than introducing Bound(Property/Field)EqualsValue, if there is a desire to avoid partially-overlapping node types.

AlekseyTs replied:
Some devs have a habit of doing type checks instead of kind checks. I think objects that we expose shouldn't implement interfaces in this fashion.

JohnHamby replied:
How about I split BoundFieldOrPropertyInitializer into BoundFieldInitializer and BoundPropertyInitializer, and eliminate BoundFieldEqualsValue and BoundPropertyEqualsValue?

genlu commented on src/Compilers/Core/Portable/Compilation/IExpression.cs:
```diff
@@ -795,31 +795,21 @@ public interface IObjectCreationExpression : IExpression
     }
 
     /// <summary>
     /// Represents an object member initializer.
```
Comment needs to be fixed as well.

JohnHamby replied:
Yes, thanks.

genlu commented on src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb:
```diff
@@ -134,7 +134,21 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                    Return New BoundFieldEqualsValue(initializer, DirectCast(Me.MemberSymbol, FieldSymbol), expressionInitializer)
+                End If
+            End If
+
```
Add `Throw ExceptionUtilities.Unreachable` at the end since this list is supposed to be exaustive

JohnHamby replied:
Yes, and change the code to a Select Case statement.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb:
```diff
@@ -1382,6 +1370,105 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+
+        Private ReadOnly Property IField As IFieldSymbol Implements IFieldInitializer.Field
+            Get
+                Return TryCast(Me.InitializedSymbols.FirstOrDefault(), IFieldSymbol)
```
It is not clear why always returning the first symbol is the right thing to do.

JohnHamby commented on src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb:
```diff
@@ -1382,6 +1370,80 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+
+        Private ReadOnly Property IInitializedProperty As IPropertySymbol Implements IPropertyInitializer.InitializedProperty
+            Get
+                Return Me.InitializedProperties.FirstOrDefault()
```
Within the VB compiler, there is logic for handling multiple properties that share an initializer in addition to multiple fields that share an initializer. Is it actually possible to write VB code in which two properties share an initializer?

JohnHamby commented on src/Compilers/VisualBasic/BasicAnalyzerDriver/VisualBasicDeclarationComputer.vb:
```diff
@@ -172,10 +172,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+            ' The As New clause itself is necessary rather than the embedded New expression, so that the
+            ' code block associated with the declaration appears as an initializer for the purposes
+            ' of executing analyzer actions.
+            Return If(asClause.IsKind(SyntaxKind.AsNewClause), asClause, Nothing)
```
This edit causes As New declarations to be treated as having member initializers by semantic models (rather than as just having object creation expressions). This could be patched within the semantic model code instead.

JohnHamby commented:
I've opened #8187 to track the issue of only the first field in an As New declaration getting through the VB semantic models. I think I've addressed all feedback, and apologize if I've missed something.

Can anyone sign off?

genlu commented:
:+1:

cston commented on src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb:
```diff
@@ -134,7 +134,20 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                    Case Else
+                        Throw ExceptionUtilities.Unreachable
+                End Select
+            End If
```
Consider merging `Select Case` with previous `Select Case`, to avoid switching twice and avoid the `TryCast`.

JohnHamby replied:
The TryCast can't be avoided, but the second switch and the additional DirectCasts can.

cston commented on src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb:
```diff
@@ -1110,37 +1104,31 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             Public ReadOnly Property IsInvalid As Boolean Implements IOperation.IsInvalid
                 Get
-                    Return Me.Value.IsInvalid OrElse Me.Setter Is Nothing
+                    Return Me.Value.IsInvalid OrElse Me.InitializedProperty Is Nothing
```
Should this include ```OrElse property.SetMethod Is Nothing```?

JohnHamby replied:
Yeah, probably. Will add.

cston commented on src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb:
```diff
@@ -1382,6 +1370,80 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+
+        Private ReadOnly Property IIsInvalid As Boolean Implements IOperation.IsInvalid
+            Get
+                Return (DirectCast(Me.Value, IExpression)).IsInvalid
```
Unnecessary parentheses.

cston commented on src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_FieldOrPropertyInitializer.vb:
```diff
@@ -12,6 +12,14 @@ Imports TypeKind = Microsoft.CodeAnalysis.TypeKind
     Partial Friend NotInheritable Class LocalRewriter
 
+        Public Overrides Function VisitFieldInitializer(node As BoundFieldInitializer) As BoundNode
+            Return VisitFieldOrPropertyInitializer(node, node.InitializedFields.Cast(Of Symbol)().ToImmutableArray())
```
```.Cast(Of FieldSymbol, Symbol)()``` which should avoid an allocation.

cston commented on src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_FieldOrPropertyInitializer.vb:
```diff
@@ -12,6 +12,14 @@ Imports TypeKind = Microsoft.CodeAnalysis.TypeKind
+        End Function
+
+        Public Overrides Function VisitPropertyInitializer(node As BoundPropertyInitializer) As BoundNode
+            Return VisitFieldOrPropertyInitializer(node, node.InitializedProperties.Cast(Of Symbol)().ToImmutableArray())
```
```.Cast(Of PropertySymbol, Symbol)()```

cston commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/FieldInitializerBindingTests.vb:
```diff
@@ -1545,7 +1545,12 @@ String
                     Dim initValueLineNumber = syntaxTree.GetLineSpan(initValueSyntax.Span).StartLinePosition.Line
                     Assert.Equal(expectedInitializer.LineNumber, initValueLineNumber)
-                    Dim fieldSymbol = boundFieldInit.InitializedSymbols.First
+                    Dim fieldSymbol As Symbol
```
Perhaps `symbol` rather than `fieldSymbol`.

JohnHamby replied:
fieldOrPropertySymbol, maybe?

cston commented on src/Compilers/VisualBasic/Test/Semantic/Semantics/FieldInitializerBindingTests.vb:
```diff
@@ -1536,7 +1536,7 @@ String
                     i += 1
-                    Assert.Equal(BoundKind.FieldOrPropertyInitializer, boundInit.Kind)
+                    Assert.[True](boundInit.Kind = BoundKind.FieldInitializer OrElse boundInit.Kind = BoundKind.PropertyInitializer)
                     Dim boundFieldInit = DirectCast(boundInit, BoundFieldOrPropertyInitializer)
```
Perhaps `boundInit` rather than `boundFieldInit`.

JohnHamby replied:
There's already a variable by that name.

cston commented:
LGTM other than minor comments.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_Initializers.vb:
```diff
@@ -178,13 +178,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                             End If
 
-                            initializerBinder.BindFieldInitializer(initializer.FieldsOrProperties,
+                            initializerBinder.BindFieldInitializer(initializer.FieldsOrProperties.Cast(Of FieldSymbol).ToImmutableArray(),
```
Is this using IEnumerable.Cast? Consider using ToDowncastedImmutable helper instead.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_Initializers.vb:
```diff
@@ -178,13 +178,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                         End If
                     Else
-                        initializerBinder.BindPropertyInitializer(initializer.FieldsOrProperties,
+                        initializerBinder.BindPropertyInitializer(initializer.FieldsOrProperties.Cast(Of PropertySymbol).ToImmutableArray(),
```
Is this using IEnumerable.Cast? Consider using ToDowncastedImmutable helper instead.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/Expression.cs:
```diff
@@ -484,46 +484,44 @@ ImmutableArray<IMemberInitializer> IObjectCreationExpression.MemberInitializers
+            public IFieldSymbol InitializedField { get; }
 
-            MemberInitializerKind IMemberInitializer.MemberInitializerKind => MemberInitializerKind.Field;
+            public ImmutableArray<IFieldSymbol> InitializedFields => ImmutableArray.Create(this.InitializedField);
```
Would it make sense to cache the array instead and return the same array every time. InitializedField property can simply return first element from the array.

AlekseyTs commented on src/Compilers/VisualBasic/BasicAnalyzerDriver/VisualBasicDeclarationComputer.vb:
```diff
@@ -172,10 +172,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-            Return If(asClause.IsKind(SyntaxKind.AsNewClause),
-                CType(asClause, AsNewClauseSyntax).NewExpression,
-                Nothing)
+        Private Shared Function GetAsNewClauseIntializer(asClause As AsClauseSyntax) As SyntaxNode
```
Consider using specific return type.
